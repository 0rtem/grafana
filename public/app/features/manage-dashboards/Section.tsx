import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { DashboardSection } from 'app/types';
import FormSwitch from 'app/core/components/FormSwitch/FormSwitch';
import SectionItem from './SectionItem';
import { collapseSection, loadSectionItems, setSectionSelected } from './state/actions';

export interface Props {
  section: DashboardSection;
  loadSectionItems: typeof loadSectionItems;
  collapseSection: typeof collapseSection;
  setSectionSelected: typeof setSectionSelected;
}

export class Section extends PureComponent<Props> {
  toggleFolder = () => {
    const { section, loadSectionItems, collapseSection } = this.props;

    if (section.expanded) {
      collapseSection(section.id);
    } else {
      loadSectionItems(section.id);
    }
  };

  toggleSectionSelected = event => {
    event.stopPropagation();

    const { section } = this.props;

    this.props.setSectionSelected(section.id);
  };

  render() {
    const { section } = this.props;

    return (
      <div className="search-section">
        <div
          className={`search-section__header pointer ${section.checked ? 'selected' : ''}`}
          onClick={this.toggleFolder}
        >
          <FormSwitch
            label=""
            onChange={event => this.toggleSectionSelected(event)}
            checked={section.checked}
            switchClass="gf-form-switch--transparent gf-form-switch--search-result__section"
          />
          <i className={`search-section__header__icon ${section.icon}`} />
          <span className="search-section__header__text">{section.title}</span>
          {section.url && (
            <a href={section.url} className="search-section__header__link">
              <i className="fa fa-cog" />
            </a>
          )}
          {section.expanded && <i className="fa fa-angle-down search-section__header__toggle" />}
          {!section.expanded && <i className="fa fa-angle-right search-section__header__toggle" />}
        </div>
        {section.hideHeader && <div className="search-section__header" />}
        {section.expanded &&
          section.items.map((item, index) => {
            return <SectionItem item={item} key={`${item.id}-${index}`} />;
          })}
      </div>
    );
  }
}

function mapStateToProps() {
  return {};
}

const mapDispatchToProps = {
  collapseSection,
  loadSectionItems,
  setSectionSelected,
};

export default connect(mapStateToProps, mapDispatchToProps)(Section);
