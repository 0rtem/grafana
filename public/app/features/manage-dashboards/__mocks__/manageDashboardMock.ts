import { DashboardSection, DashboardSectionItem } from '../../../types';

export const getMockSections = (amount: number): DashboardSection[] => {
  const sections: DashboardSection[] = [];

  for (let i = 1; i <= amount; i++) {
    sections.push({
      id: i,
      uid: `asdf${i}`,
      title: 'Such a cool folder',
      expanded: false,
      items: [] as DashboardSectionItem[],
      url: `some/url/${i}`,
      icon: 'cool-icon',
      score: i,
      hideHeader: false,
      checked: false,
    });
  }

  return sections;
};

export const getMockSectionsWithItems = (amount: number): DashboardSection[] => {
  return getMockSections(amount).map(section => {
    section.items = getMockSectionItems(amount);

    return section;
  });
};

export const getMockSection = (): DashboardSection => {
  return {
    id: 1,
    uid: 'asdf1',
    title: 'Such a cool folder',
    expanded: false,
    items: [] as DashboardSectionItem[],
    url: 'some/url/1',
    icon: 'cool-icon',
    score: 1,
    hideHeader: false,
    checked: false,
  };
};

export const getMockSectionItems = (amount: number): DashboardSectionItem[] => {
  const items: DashboardSectionItem[] = [];

  for (let i = 0; i <= amount; i++) {
    items.push({
      id: i,
      uid: `id-${i}`,
      title: `item-${i}`,
      uri: '',
      url: `some/url/${i}`,
      type: '',
      tags: ['blue', 'green'],
      isStarred: false,
      folderId: 1,
      folderUid: 'folder-1',
      folderTitle: 'folderur',
      folderUrl: 'some/folderUrl/1',
      checked: false,
    });
  }

  return items;
};
