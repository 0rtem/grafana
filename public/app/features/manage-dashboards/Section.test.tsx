import React from 'react';
import { shallow } from 'enzyme';
import { Props, Section } from './Section';
import { getMockSection, getMockSectionItems } from './__mocks__/manageDashboardMock';

const setup = (propOverrides?: object) => {
  const props: Props = {
    section: getMockSection(),
    loadSectionItems: jest.fn(),
    collapseSection: jest.fn(),
    setSectionSelected: jest.fn(),
  };

  Object.assign(props, propOverrides);

  const wrapper = shallow(<Section {...props} />);
  const instance = wrapper.instance() as Section;

  return {
    wrapper,
    instance,
  };
};

describe('Render', () => {
  it('should render component', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render items', () => {
    const { wrapper } = setup({
      section: {
        ...getMockSection(),
        items: getMockSectionItems(5),
        expanded: true,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Functions', () => {
  describe('Toggle folder expand', () => {
    it('should expand folder', () => {
      const { instance } = setup();

      instance.toggleFolder();

      expect(instance.props.loadSectionItems).toHaveBeenCalled();
    });

    it('should collapse folder', () => {
      const { instance } = setup({ section: { ...getMockSection(), expanded: true } });

      instance.toggleFolder();

      expect(instance.props.collapseSection).toHaveBeenCalled();
    });
  });
});
