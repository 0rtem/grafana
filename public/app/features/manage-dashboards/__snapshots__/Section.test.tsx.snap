// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="search-section"
>
  <div
    className="search-section__header pointer "
    onClick={[Function]}
  >
    <FormSwitch
      label=""
      onChange={[Function]}
      switchClass="gf-form-switch--transparent gf-form-switch--search-result__section"
    />
    <i
      className="search-section__header__icon cool-icon"
    />
    <span
      className="search-section__header__text"
    >
      Such a cool folder
    </span>
    <a
      className="search-section__header__link"
      href="some/url/1"
    >
      <i
        className="fa fa-cog"
      />
    </a>
    <i
      className="fa fa-angle-right search-section__header__toggle"
    />
  </div>
</div>
`;

exports[`Render should render items 1`] = `
<div
  className="search-section"
>
  <div
    className="search-section__header pointer "
    onClick={[Function]}
  >
    <FormSwitch
      label=""
      onChange={[Function]}
      switchClass="gf-form-switch--transparent gf-form-switch--search-result__section"
    />
    <i
      className="search-section__header__icon cool-icon"
    />
    <span
      className="search-section__header__text"
    >
      Such a cool folder
    </span>
    <a
      className="search-section__header__link"
      href="some/url/1"
    >
      <i
        className="fa fa-cog"
      />
    </a>
    <i
      className="fa fa-angle-down search-section__header__toggle"
    />
  </div>
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 0,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-0",
        "type": "",
        "uid": "id-0",
        "uri": "",
        "url": "some/url/0",
      }
    }
    key="0-0"
  />
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 1,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-1",
        "type": "",
        "uid": "id-1",
        "uri": "",
        "url": "some/url/1",
      }
    }
    key="1-1"
  />
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 2,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-2",
        "type": "",
        "uid": "id-2",
        "uri": "",
        "url": "some/url/2",
      }
    }
    key="2-2"
  />
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 3,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-3",
        "type": "",
        "uid": "id-3",
        "uri": "",
        "url": "some/url/3",
      }
    }
    key="3-3"
  />
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 4,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-4",
        "type": "",
        "uid": "id-4",
        "uri": "",
        "url": "some/url/4",
      }
    }
    key="4-4"
  />
  <Connect(SectionItem)
    item={
      Object {
        "folderId": 1,
        "folderTitle": "folderur",
        "folderUid": "folder-1",
        "folderUrl": "some/folderUrl/1",
        "id": 5,
        "isStarred": false,
        "tags": Array [
          "blue",
          "green",
        ],
        "title": "item-5",
        "type": "",
        "uid": "id-5",
        "uri": "",
        "url": "some/url/5",
      }
    }
    key="5-5"
  />
</div>
`;
